# Entity Matching using Neo4j - Rare ngrams

_Salomon Tetelepta, May 4th 2024_
* Notebook to improve the model by enrich the graph
    * detect model numbers based on ngram rarity

### Install dependencies

In [1]:
!pip install neo4j python-dotenv langchain-community --quiet

%load_ext watermark
%watermark -p neo4j

neo4j: 5.17.0



### Imports

In [2]:
from dotenv import load_dotenv, find_dotenv, dotenv_values
from langchain_community.graphs import Neo4jGraph
from pathlib import Path
from sklearn.manifold import TSNE
from sklearn.metrics import PrecisionRecallDisplay
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix, f1_score
from sklearn.metrics import precision_recall_fscore_support
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from wordfreq import lossy_tokenize, tokenize, word_frequency

import json
import matplotlib.pyplot as plt
import neo4j
import numpy as np

import os
import pandas as pd
import pickle
import re

### Settings

In [3]:
# path settings
project_path = Path(os.getcwd()).parent
data_path = project_path / "data"
output_path = project_path / "output"

database = "abt-buy"

# load env settings
load_dotenv(project_path / ".env")

# reproducability
np.random.seed(42)

### 1. Load Data

In [4]:
os.listdir(data_path / 'abt-buy' / 'record_descriptions')

['2_buy.csv', '1_abt.csv']

In [5]:
# abt and buy records
df_abt = pd.read_csv(data_path / 'abt-buy' / 'record_descriptions' / '1_abt.csv', encoding='unicode_escape')
df_buy = pd.read_csv(data_path / 'abt-buy' / 'record_descriptions' / '2_buy.csv')

# matches - train and validation set
df_train = pd.read_csv(data_path / 'abt-buy' / 'gs_train.csv')
df_val = pd.read_csv(data_path / 'abt-buy' / 'gs_val.csv')
df_test = pd.read_csv(data_path / 'abt-buy' / 'gs_test.csv')

# merge records with matches
df_abt_merged = df_abt.merge(df_train, left_on='subject_id', right_on='source_id', how='right')
df_train_merged = df_buy.merge(df_abt_merged, left_on='subject_id', right_on='target_id', how='right')

df_train_merged.head(3)

subject_id_x                                             name_x  \
0     207390654  Sony Handycam HDR-SR10 High Definition Digital...   
1     208085180                Pioneer DEH-2000MP Car Audio Player   
2      90125786    Sanus Wall/Ceiling Speaker Mount - WMS3S SILVER   

                                       description_x manufacturer  price_x  \
0                             16:9 - 2.7' Hybrid LCD         Sony    549.0   
1  CD-RW - CD-Text, MP3, WMA, WAV - LCD - 4 - 200...      Pioneer     84.0   
2                                     Plastic - 8 lb        Sanus      NaN   

   subject_id_y                                             name_y  \
0         33161  Sony High Definition HDV Handycam Camcorder - ...   
1         36258              D-Link Broadband Cable Modem - DCM202   
2         17417  Sanus 13' - 30' VisionMount Flat Panel TV Silv...   

                                       description_y  price_y  source_id  \
0  Sony High Definition HDV Handycam Camcorder - ...      NaN      33161   
1  D-Link Broadband Cable Modem - DCM202/ DOCSIS ...    79.00      36258   
2  Sanus 13' - 30' VisionMount Flat Panel TV Silv...    39.99      17417   

   target_id  matching  
0  207390654     False  
1  208085180     False  
2   90125786     False

### Connect to Neo4j

In [6]:
# connect to Neo4j
graph = Neo4jGraph(
    url=os.getenv('NEO4J_URL'),
    username=os.getenv('NEO4J_USER'),
    password=os.getenv('NEO4J_PASS')
)

# create database if does not exist
graph._database = "system"
query = f"CREATE DATABASE `{database}` IF NOT EXISTS"
graph.query(query)

# change to target database
graph._database = database
print("database:", graph._database)

# check nr nodes in the graph
graph.query("MATCH (n) RETURN count(n)")

database: abt-buy


[{'count(n)': 612119}]

### Create Ngram dataset

In [7]:
import nltk
from nltk.corpus import words
from nltk.util import ngrams
from nltk.tokenize import word_tokenize

from itertools import islice, tee
from collections import defaultdict

In [8]:
%%time
# Wall time: 11.2 s

run_cell = False
if run_cell == True:
    nltk.download('words')

CPU times: user 5 µs, sys: 2 µs, total: 7 µs
Wall time: 11 µs


In [9]:
def get_ngrams(s, n):
    ngrams = list(zip(*(islice(seq, index, None) for index, seq in enumerate(tee(s, n)))))
    return ["".join(n) for n in ngrams]

In [10]:
nr_words = None
list_words = [w.lower() for w in words.words()[:nr_words]]

In [11]:
%%time
# Wall time: 5.99 s

nr_words = None
list_words = [w.lower() for w in words.words()[:nr_words]]

output = []
for n in np.arange(3, 7):
    dict_ngrams = defaultdict(int)
    dict_ngrams
    for w in list_words:
        for ngram in get_ngrams(w, n=n):
            dict_ngrams[ngram] += 1

    df_ngrams = pd.DataFrame([dict_ngrams]).T.sort_values(0, ascending=False).rename(columns={0: 'count'})
    df_ngrams['n'] = n
    output.append(df_ngrams)
df_ngrams_all = pd.concat(output)

CPU times: user 6.02 s, sys: 110 ms, total: 6.13 s
Wall time: 6.14 s


### Save to CSV

In [12]:
df_ngrams_all.reset_index(names='ngram').to_csv(output_path / 'ngrams.csv', index=False)
df_ngrams_all = pd.read_csv(output_path / 'ngrams.csv')

### Add ngrams to Graph

In [13]:
df_ngrams_all.head()

ngram  count  n
0   ess  10976  3
1   ion   9704  3
2   ter   9546  3
3   ati   9473  3
4   ing   8951  3

#### Add unique constraint

In [14]:
query = "CREATE CONSTRAINT ngram_value IF NOT EXISTS FOR (ngram:Ngram) REQUIRE ngram.value IS UNIQUE"
graph.query(query)

[]

#### brute force ngrams to graph

In [15]:
%%time
# Wall time: 2h 42min 10s

run_cell = False
if run_cell:
    for i, (idx, row) in enumerate(df_ngrams_all.iterrows()):
        if i % 50_000 == 0:
            print(f"{i} / {len(df_ngrams_all):,}")
        query = f"MERGE (ngram:Ngram {{value: '{ row['ngram']}', count: {row['count']} }})"
        graph.query(query)

CPU times: user 6 µs, sys: 3 µs, total: 9 µs
Wall time: 17.9 µs


### Split WordLower nodes into Ngrams

* Create relationships `(:WordLower)-[:HAS_[N]GRAM]->(:Ngram)`

In [16]:
query = """MATCH (w:WordLower) RETURN w.value AS wordlower"""
results = graph.query(query)
df_results = pd.DataFrame(results)
df_results.head()

wordlower
0       sony
1  turntable
2          -
3   pslx350h
4  pslx350h/

In [18]:
for n in np.arange(3, 7):
    df_results.loc[:, f'{n}gram'] = df_results['wordlower'].apply(lambda x: get_ngrams(x, n))
    
df_results

wordlower                                     3gram  \
0           sony                                [son, ony]   
1      turntable       [tur, urn, rnt, nta, tab, abl, ble]   
2              -                                        []   
3       pslx350h            [psl, slx, lx3, x35, 350, 50h]   
4      pslx350h/       [psl, slx, lx3, x35, 350, 50h, 0h/]   
...          ...                                       ...   
9711    mb942z/a            [mb9, b94, 942, 42z, 2z/, z/a]   
9712    pack-int            [pac, ack, ck-, k-i, -in, int]   
9713    mb943z/a            [mb9, b94, 943, 43z, 3z/, z/a]   
9714        mate                                [mat, ate]   
9715  iph3gc-bcf  [iph, ph3, h3g, 3gc, gc-, c-b, -bc, bcf]   

                                           4gram  \
0                                         [sony]   
1           [turn, urnt, rnta, ntab, tabl, able]   
2                                             []   
3                 [pslx, slx3, lx35, x350, 350h]   
4           [pslx, slx3, lx35, x350, 350h, 50h/]   
...                                          ...   
9711              [mb94, b942, 942z, 42z/, 2z/a]   
9712              [pack, ack-, ck-i, k-in, -int]   
9713              [mb94, b943, 943z, 43z/, 3z/a]   
9714                                      [mate]   
9715  [iph3, ph3g, h3gc, 3gc-, gc-b, c-bc, -bcf]   

                                           5gram  \
0                                             []   
1            [turnt, urnta, rntab, ntabl, table]   
2                                             []   
3                   [pslx3, slx35, lx350, x350h]   
4            [pslx3, slx35, lx350, x350h, 350h/]   
...                                          ...   
9711                [mb942, b942z, 942z/, 42z/a]   
9712                [pack-, ack-i, ck-in, k-int]   
9713                [mb943, b943z, 943z/, 43z/a]   
9714                                          []   
9715  [iph3g, ph3gc, h3gc-, 3gc-b, gc-bc, c-bcf]   

                                         6gram  
0                                           []  
1             [turnta, urntab, rntabl, ntable]  
2                                           []  
3                     [pslx35, slx350, lx350h]  
4             [pslx35, slx350, lx350h, x350h/]  
...                                        ...  
9711                  [mb942z, b942z/, 942z/a]  
9712                  [pack-i, ack-in, ck-int]  
9713                  [mb943z, b943z/, 943z/a]  
9714                                        []  
9715  [iph3gc, ph3gc-, h3gc-b, 3gc-bc, gc-bcf]  

[9716 rows x 5 columns]

In [19]:
%%time
# Wall time: 21min 22s

run_cell = False
if run_cell:
    for i, (idx, row) in enumerate(df_results.iterrows()):
        if i % 1000 == 0:
            print(f"{i}/{len(df_results)}")

        for n in np.arange(3, 7):
            for ngram in row[f'{n}gram']:
                value = row['wordlower'].replace("'", "\\'")
                ngram = ngram.replace("'", "\\'")

                query = f"""
                MATCH (w:WordLower) WHERE w.value = '{value}'
                MERGE (ngram:Ngram {{value: '{ngram}'}})
                MERGE (w)-[:HAS_{n}GRAM]->(ngram)
                SET w:PROCESSED
                RETURN count(ngram)
                """
                graph.query(query)

CPU times: user 4 µs, sys: 1e+03 ns, total: 5 µs
Wall time: 8.11 µs


#### Label rare ngrams

In [20]:
# first, label ngrams that have a count attribute, which means they appeared in the vocabulary
query = "MATCH (n:Ngram) WHERE n.count > 0 SET n:Vocab"
graph.query(query)

# next, label rare ngrams, the ones that do not appear in the vocabulary
query = "MATCH (n:Ngram) WHERE NOT n:Vocab SET n:NoVocab"
graph.query(query)

[]

#### Examples

<h4>Can we find model IDs and other reference ids by checking for rare ngrams that words have in common?</h4>
<img src="../images/7.1-example-sharing-multiple-rare-3grams.jpg" width="600">

<h4>Example of matches that share a rare ngram </h4>
<img src="../images/7.2-example-sharing-multiple-rare-3grams.jpg" width="600">

<pre>MATCH p=(ngram:NoVocab)<-[]-(:WordLower)<-[:HAS_WORD_LOWER]-(:Name)<-[:HAS_NAME]-(i1:Item)-[:IS_MATCH]-(i2:Item)-[:HAS_NAME]->(:Name)-[:HAS_WORD_LOWER]->(:WordLower)-[]->(ngram) RETURN p LIMIT 10</pre>

<h4>Example of no matches that share a rare ngram </h4>
<img src="../images/7.2-example-sharing-multiple-rare-3grams.jpg" width="600">

<pre>MATCH p=(ngram:NoVocab)<-[]-(:WordLower)<-[:HAS_WORD_LOWER]-(:Name)<-[:HAS_NAME]-(i1:Item)-[:NO_MATCH]-(i2:Item)-[:HAS_NAME]->(:Name)-[:HAS_WORD_LOWER]->(:WordLower)-[]->(ngram) RETURN p LIMIT 10</pre>

<h4>Example of no matches that share a rare ngram </h4>
<img src="../images/7.4-example-sharing-multiple-rare-3grams.jpg" width="600">

<h4>Example of an incorrect match where there is a crucial difference</h4>
<img src="../images/7.5-example-crucial-difference.jpg" width="600">

<h4>Example of an incorrect match where there is a common 6gram</h4>
<img src="../images/7.6-example-common-6gram.jpg" width="600">

### Evaluate query that takes into account sharing rare ngrams

#### Create relations between items that share rare ngrams

* New model extends the previous model using a UNION ALL
* Above the threshold of 6 additional rule is to check for shared ngrams of a certain size {4, 5, 6}

In [36]:
results = []

In [37]:
%%time

threshold_low = 6
threshold_high = 47
ngram_size = 6

for ngram_size in [4, 5, 6]:
    query = f"""
    MATCH p1=(i1:Item {{source: 'abt'}})-[:HAS_NAME]->(n1:Name)-[:HAS_WORD_LOWER]->(w:WordLower)<-[:HAS_WORD_LOWER]-(n2:Name)<-[:HAS_NAME]-(i2:Item {{source: 'buy'}})
    WHERE n1 <> n2
    AND w.name_degree < {threshold_low}
    RETURN i1.subject_id, i2.subject_id
    UNION ALL
    MATCH p1=(i1:Item {{source: 'abt'}})-[:HAS_NAME]->(n1:Name)-[:HAS_WORD_LOWER]->(w:WordLower)<-[:HAS_WORD_LOWER]-(n2:Name)<-[:HAS_NAME]-(i2:Item {{source: 'buy'}})
    WHERE n1 <> n2
    AND w.name_degree >= {threshold_low}
    AND w.name_degree < {threshold_high}
    AND (n1)-[:HAS_WORD_LOWER]->(:WordLower)-[:HAS_{ngram_size}GRAM]->(:NoVocab)<-[:HAS_{ngram_size}GRAM]-(:WordLower)<-[:HAS_WORD_LOWER]-(n2)
    RETURN i1.subject_id, i2.subject_id
    """

    #print(query)
    df_p = pd.DataFrame(graph.query(query))

    if len(df_p) > 0:

        df_test_p = df_test.merge(df_p, left_on=['source_id', 'target_id'], right_on=['i1.subject_id', 'i2.subject_id'], how='left')
        df_test_p['p'] = df_test_p['i1.subject_id'] > 0

        prec, recall, fscore, support = precision_recall_fscore_support(df_test_p['matching'], df_test_p['p'], average='binary')

        # store errors
        cond_p1 = df_test_p['p'] == True
        cond_y1 = df_test_p['matching'] == True

        df_tp = df_test_p[cond_p1 & cond_y1]
        df_fp = df_test_p[cond_p1 & ~cond_y1]
        df_tn = df_test_p[~cond_p1 & ~cond_y1]
        df_fn = df_test_p[~cond_p1 & cond_y1]

        tp = len(df_tp)
        fp = len(df_fp)
        fn = len(df_fn)
        tn = len(df_tn)

        results.append({'model': f'shared_word_lower_threshold_{ngram_size}gram', 'threshold': (threshold_low, threshold_high), 'prec': prec, 'recall': recall, 'fscore': fscore, 'evaluated_on': 'testset', 'tp': tp, 'fp': fp, 'fn': fn, 'tn': tn})

    df_results = pd.DataFrame(results).sort_values('fscore', ascending=False).reset_index(drop=True)
    display(df_results.head())

model threshold      prec    recall    fscore  \
0  shared_word_lower_threshold_4gram   (6, 47)  0.669782  0.903361  0.769231   

  evaluated_on   tp   fp  fn   tn  
0      testset  215  106  23  543

model threshold      prec    recall    fscore  \
0  shared_word_lower_threshold_5gram   (6, 47)  0.727273  0.880383  0.796537   
1  shared_word_lower_threshold_4gram   (6, 47)  0.669782  0.903361  0.769231   

  evaluated_on   tp   fp  fn   tn  
0      testset  184   69  25  559  
1      testset  215  106  23  543

model threshold      prec    recall    fscore  \
0  shared_word_lower_threshold_6gram   (6, 47)  0.801980  0.843750  0.822335   
1  shared_word_lower_threshold_5gram   (6, 47)  0.727273  0.880383  0.796537   
2  shared_word_lower_threshold_4gram   (6, 47)  0.669782  0.903361  0.769231   

  evaluated_on   tp   fp  fn   tn  
0      testset  162   40  30  574  
1      testset  184   69  25  559  
2      testset  215  106  23  543

CPU times: user 537 ms, sys: 66.9 ms, total: 604 ms
Wall time: 15.4 s


In [40]:
df_results_all = pd.read_csv(output_path / 'results.csv')
df_results_all = pd.concat([df_results, df_results_all]).fillna(0).sort_values('fscore', ascending=False)
df_results_all = df_results_all.astype({'tp': 'int', 'fp': 'int', 'fn': 'int', 'tn': 'int'}).reset_index(drop=True)
df_results_all.head()

model threshold      prec    recall    fscore  \
0  shared_word_lower_threshold_6gram   (6, 47)  0.801980  0.843750  0.822335   
1  shared_word_lower_threshold_5gram   (6, 47)  0.727273  0.880383  0.796537   
2        shared_word_lower_threshold         6  0.896226  0.673759  0.769231   
3  shared_word_lower_threshold_4gram   (6, 47)  0.669782  0.903361  0.769231   
4        shared_word_lower_threshold        10  0.754601  0.763975  0.759259   

  evaluated_on   tp   fp  fn   tn  
0      testset  162   40  30  574  
1      testset  184   69  25  559  
2      testset    0    0   0    0  
3      testset  215  106  23  543  
4      testset    0    0   0    0

### Results

* sharing 6-grams gives best results
* precision went down from `0.896226` to `0.801980` (`-9.4%pt`)
* recall went up from `0.673759` to `0.843750` (`+17.0%pt`)
* overall f-score went up from `0.769231` to `0.822335` (`+5.3%pt`)

### Idea:

_crucial differences_
* For words that share a large ngram, what makes the diffference, and is it crucial?
* In the case `pslx350h` vs `ps-lx350h` 
    * largest non-overlapping ngrams: `ps` (2-gram) and `lx350h` (6-gram) -> specific
    * difference: `-` -> common

_Manufacturers_
* Some of the items share words or non-vocabulary ngrams that are manufacturer names.

_Model rules as votes_
* Rather than creating increasingly complex queries to capture more subtle rules, we might model rules as votes for a match or no-match
    * Votes can be positive (ie. sharing rare ngrams) or negative (ie. crucial differences)
    * Votes should have some kind of weight (ie. approximate conditional probability)